### DISCUSSION TITLE
render pdf and save to disk

### DISCUSSION THREAD

**Discussion by teafmartins** [serenity-is/Serenity#5351] [Help]
Hello,

I need to render PDF  and save it to disk. 
It works fine in localhost but not on production server (domain server), I get "Object reference not set to an instance of an object." error.

I'm using this endpoint:

        [HttpPost]
        public StringResponse SaveToDisk(IDbConnection connection, string Id)
        {
            var result = new StringResponse();
            result.Value = "";
            var reportOptions = new Dictionary<string, string>
                {
                    { "Id", "9999" },
                };
            var opt = JsonConvert.SerializeObject(reportOptions);
            string key = "Process.Report1";

            var reportInfo = ReportRegistry.GetReport(key);

            var report = (IReport)JsonConvert.DeserializeObject(opt.TrimToNull() ?? "{}",
                        reportInfo.Type, JsonSettings.Tolerant);
            byte[] renderedBytes = null;

            renderedBytes = new ReportController().RenderAsPdf(report, key, opt, ControllerContext);
            try
            {
                var filenamePath = string.Format(@"{0}\\Files\\{1}\\", Server.MapPath("~"), Id);
                var filename = string.Format(@"{0}{1}.PDF", filenamePath, Id);
                System.IO.Directory.CreateDirectory(filenamePath);
                System.IO.File.WriteAllBytes(filename, renderedBytes);
                result.Value = "OK";
            }
            catch (Exception)
            {
            }

            return result;
        }

Why it works fine in localhost but not on domain server?!
Thanks

**Comment by patriciomh1965** [comment]
Dear, I can give you an idea to download a file, in this case it is an EXCEL spreadsheet.

**Grid :**

namespace SegurosMAPC.BaseDatos {

    @Serenity.Decorators.registerClass()
    export class CuadraturaFiltrosGrid extends Serenity.EntityGrid<CuadraturaFiltrosRow, any> {
        protected getColumnsKey() { return 'BaseDatos.CuadraturaFiltros'; }
        protected getDialogType() { return CuadraturaFiltrosDialog; }
        protected getIdProperty() { return CuadraturaFiltrosRow.idProperty; }
        protected getInsertPermission() { return CuadraturaFiltrosRow.insertPermission; }
        protected getLocalTextPrefix() { return CuadraturaFiltrosRow.localTextPrefix; }
        protected getService() { return CuadraturaFiltrosService.baseUrl; }

        constructor(container: JQuery) {
            super(container);
        }




        //  OPERACIONES LINEA
        protected getColumns() {
            var columns = super.getColumns();
            
            columns.splice(0, 0, {
                field: 'Print Invoice',
                name: '',
                format: ctx => '<a class="inline-action print-invoice" title="Permite crear la planilla EXCEL">' +
                    '<i class="fa fa-calendar-plus-o text-black"></i></a>',                 
                width: 24,
                minWidth: 24,
                maxWidth: 24
            });

            columns.splice(1, 0, {
                field: 'Descargar Archivo',
                name: '',
                format: ctx => '<a class="inline-action download-file" title="Descargar planilla EXCEL">' +
                    '<i class="fa fa-file-excel-o text-green"></i></a>',
                width: 24,
                minWidth: 24,
                maxWidth: 24
            });


            return columns;
        }

        protected onClick(e: JQueryEventObject, row: number, cell: number) {
            super.onClick(e, row, cell);

            if (e.isDefaultPrevented())
                return;

            var item = this.itemAt(row);
            var reporteestado = item.ReporteEstado;
            var target = $(e.target);

            // if user clicks "i" element, e.g. icon
            if (target.parent().hasClass('inline-action'))
                target = target.parent();

            if (target.hasClass('inline-action')) {
                e.preventDefault();
                
                if (target.hasClass('print-invoice')) {
                    // 
                    if (reporteestado == 0 ) { Q.notifyWarning("Debe generar los datos para planilla EXCEL"); return; }  //                    
                    if (reporteestado == 1 || reporteestado == 4 )
                    { Q.notifyWarning("No puede crear la planilla Excel  ... aun no termina"); return; }  //                    
                    if (reporteestado == 5 ) { Q.notifyWarning("Planilla Excel  ya se encuentra terminada ... "); return; }  //                    
                    var c = <CuadraturaFiltrosRow>{};
                    CuadraturaFiltrosService.ReporteExcel({
                        EntityId: item.Id,
                        Entity: c
                    }, response => {
                        Q.notifySuccess("Se esta generando planilla EXCEL", "Planilla EXCEL", { timeOut: 5000 });
                        this.refresh();
                        this.element.triggerHandler('');
                    });
                }

                //
                if (target.hasClass('download-file')) {
                    if (reporteestado == 0) { Q.notifyWarning("Debe generar los datos para planilla EXCEL"); return; }  //      
                    if (reporteestado == 1 || reporteestado == 2 || reporteestado == 3 || reporteestado == 4 )
                       { Q.notifyWarning("No puede descargar la planilla Excel  ... no se encuentra disponible"); return; }  //   
                    let ServiceEndpoint = this.getService() + "/ExportFile";
                    Q.postToService({ service: ServiceEndpoint, request: { EntityId: item.Id }, target: '_blank' });
                }
            }
        }
        //
    }
}

**EndPoint :** 

namespace SegurosMAPC.BaseDatos.Endpoints
{
    using Serenity;
    using Serenity.Data;
    using Serenity.Services;
    using System.Data;
    using System.Net;
    using System.Web.Mvc;
    using MyRepository = Repositories.CuadraturaFiltrosRepository;
    using MyRow = Entities.CuadraturaFiltrosRow;

    //[Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
    [RoutePrefix("Services/BaseDatos/CuadraturaFiltros"), Route("{action}")]
    // [ConnectionKey(typeof(MyRow)), ServiceAuthorize(typeof(MyRow))]
    [ConnectionKey(typeof(MyRow))]

    public class CuadraturaFiltrosController : ServiceEndpoint
    {
        [HttpPost, AuthorizeCreate(typeof(MyRow))]
        public SaveResponse Create(IUnitOfWork uow, SaveRequest<MyRow> request)
        {
            return new MyRepository().Create(uow, request);
        }

        [HttpPost, AuthorizeUpdate(typeof(MyRow))]
        public SaveResponse Update(IUnitOfWork uow, SaveRequest<MyRow> request)
        {
            return new MyRepository().Update(uow, request);
        }
 
        [HttpPost, AuthorizeDelete(typeof(MyRow))]
        public DeleteResponse Delete(IUnitOfWork uow, DeleteRequest request)
        {
            return new MyRepository().Delete(uow, request);
        }

        [HttpPost]
        public RetrieveResponse<MyRow> Retrieve(IDbConnection connection, RetrieveRequest request)
        {
            return new MyRepository().Retrieve(connection, request);
        }

        [HttpPost]
        public ListResponse<MyRow> List(IDbConnection connection, ListRequest request)
        {
            return new MyRepository().List(connection, request);
        }


        [HttpPost, AuthorizeUpdate(typeof(MyRow))]
        public SaveResponse ReporteExcel(IUnitOfWork uow, SaveRequest<MyRow> request)
        {            
            return new MyRepository().ReporteExcel(uow, request); 
        }

        [HttpPost]       
        public ActionResult ExportFile(IDbConnection connection, RetrieveRequest request)
        {
            RetrieveResponse<MyRow> Response = new MyRepository().Retrieve(connection, request);
            var archivoexcel = Response.Entity.Path + Response.Entity.NombreArchivo;
            var output = File(archivoexcel, System.Net.Mime.MediaTypeNames.Application.Octet, Response.Entity.NombreArchivo);
            return output;
        }

    }
}


Hope something helps you.

**Reply by teafmartins** [reply]
Hello @patriciomh1965 
Thanks for your code. May be I didn't explaint it very weel, I want to **render pdf** and **save it on the site server**.

**Comment by VictorTomaili** [comment]
Ensure about do you have wkhtmltopdf executable file in your server.

**Comment by teafmartins** [comment]
Hi @VictorTomaili, Yes, wkhtmltopdf is on server. Any other idea?

**Reply by VictorTomaili** [reply]
No clue, because error what you give is too much generic. You need share your code and exact exception with stack trace.
